### ISSUE TITLE
box-android-sdk. BoxAuthentication.AuthListener#onLoggedOut is not called properly

### ISSUE THREAD

**Issue by MaksimDmitriev** [box/box-android-sdk#280]
Hello.

My device is Nexus 5 with Android 6.0.1. I installed box-content-sample from https://github.com/box/box-android-sdk to my device.

I logged in to my developer's account on my device. When I pressed "Log out" on the options menu, _sometimes_ the method onLoggedOut wasn't called.

In 100 % cases onLoggedOut wasn't called when I pressed "Log out" and then pressed the Home button immediately. However, here the app wasn't killed by the system. So that's not the reason why onLoggedOut wasn't called.

Best Regards,

MAKSIM

**Comment by doncung** [comment]
Thanks, we'll look into this. The logic here is based on a weak reference so if the session is lost or the activity is destroyed that could account for the onLoggedOut call not happening. As a simple test can you add this call somewhere in your code and see if the logged out statement is printed?

            BoxAuthentication.getInstance().addListener(new BoxAuthentication.AuthListener() {
                @Override
                public void onRefreshed(BoxAuthentication.BoxAuthenticationInfo info) {
                    
                }

                @Override
                public void onAuthCreated(BoxAuthentication.BoxAuthenticationInfo info) {

                }

                @Override
                public void onAuthFailure(BoxAuthentication.BoxAuthenticationInfo info, Exception ex) {

                }

                @Override
                public void onLoggedOut(BoxAuthentication.BoxAuthenticationInfo info, Exception ex) {
                    System.out.println("onLoggedOut");
                }
            });

**Comment by MaksimDmitriev** [comment]
> As a simple test can you add this call somewhere in your code and see if the logged out statement is printed?

I did. Besides I forked the box Android example to my account and moved all the login, logout, and upload logic to a retained fragment. Here is the link to my changes: https://github.com/MaksimDmitriev/box-android-sdk/commit/d94b690f0d9edce3a0a7a42b22c712cf5880b71f

Once `BoxAuthentication onLoggedOut` was printed, and `onLoggedOut` wasn't. In the other tests both log messages were printed. 

Hope this helps!
MAKSIM

**Comment by doncung** [comment]
Thanks Maksim that's very helpful. Off the top my head it seems that either the session is being removed as a listener entirely due to some kind of GC, or the session is not relaying the information to its listener. I'll schedule this for further investigation.

**Comment by MaksimDmitriev** [comment]
Thank you for replying so quickly here on Github. When do you think your team will fix it?

**Comment by doncung** [comment]
I will try to schedule it for this sprint so hopefully should have a fix on
maven by next week.

**Comment by doncung** [comment]
Are you sure you are seeing the BoxAuthentication onLoggedOut call and NOT the onLoggedOut mActivityName=MainActivity when you were seeing only one? I have been trying to use your logic to reproduce this, and I was able to reproduce seeing just "onLoggedOut mActivityName=MainActivity".

Doing some debugging this behavior seems correct for the most part assuming the activity in which these sessions that act as the listener has not been GCed. In my case the new AuthListener added to BoxAuthentication right before logout was GCed (I added a Sysem.gc()) due to there being no hard references of it kept. In your original post are you certain your activities were not GCed after hitting home? Do you have any dev options set to kill activities immediately?

One thing you may want to try is creating an auth listener and holding onto it inside MyApp . That should stop it from being GCed until the application dies.

**Comment by MaksimDmitriev** [comment]
@doncung, the message `onLoggedOut mActivityName=MainActivity` was introduced in commit `96c5cf3` for [issue 285.](https://github.com/box/box-android-sdk/issues/285)

Please clone [my repository](https://github.com/MaksimDmitriev/box-android-sdk) and switch to a new branch like this:

` git checkout -b issue_280 d94b690`

Then log in to you Box account, press Log out, and then press the Home button immediately. You should see 

```
06-15 11:07:23.211 19800-22365/com.box.androidsdk.sample D/TaskHolder: onLoggedOut
06-15 11:07:23.211 19800-22365/com.box.androidsdk.sample D/TaskHolder: BoxAuthentication onLoggedOut
```

or just 

```
06-15 11:07:23.211 19800-22365/com.box.androidsdk.sample D/TaskHolder: BoxAuthentication onLoggedOut
```

**Update 1**

I just ran the described scenario, and even `BoxAuthentication onLoggedOut` wasn't printed.

> Do you have any dev options set to kill activities immediately?

No, the option "Do not keep activities" in developer options is unchecked. Neither do I have any third-party task killers. 

My device: Samsung Galaxy Tab S2 with Android 7.0

**Comment by PreciselyAlyss** [comment]
Appears to have been addressed by [PR#288](https://github.com/box/box-android-sdk/pull/288)
